### PR TITLE
ddl,lightning: fix ingest data unexpectedly using zero commit ts

### LINKED ISSUE #48804: more than one tikv panic repeatedly and service is unavailable when run workload and add index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

tidb_enable_dist_task='on'
1、run sysbench
2、add index for one table
3、rolling restart pd

### 2. What did you expect to see? (Required)

Add index can success and no panic.

### 3. What did you see instead (Required)

tikv panic repeatedly.

```
{"container":"tikv","pod":"tc-tikv-2","level":"FATAL","namespace":"endless-ha-test-add-index-tps-4710408-1-283","log":"[lib.rs:510] ["assertion failed: !last_change_ts.is_zero()"] [backtrace=" 0: tikv_util::set_panic_hook::{{closure}}
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/lib.rs:509:18
 1: <alloc::boxed::Box<F,A> as core::ops::function::Fn>::call
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2032:9
 std::panicking::rust_panic_with_hook
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:692:13
 2: std::panicking::begin_panic_handler::{{closure}}
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:577:13
 3: std::sys_common::backtrace::__rust_end_short_backtrace
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:137:18
 4: rust_begin_unwind
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:575:5
 5: core::panicking::panic_fmt
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:65:14
 6: core::panicking::panic
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:114:5
 7: tikv::storage::txn::actions::common::next_last_change_info
 8: tikv::storage::txn::actions::prewrite::PrewriteMutation::check_for_newer_version
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/txn/actions/prewrite.rs:437:21
 9: tikv::storage::txn::actions::prewrite::prewrite
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/txn/actions/prewrite.rs:86:10
 10: tikv::storage::txn::commands::prewrite::Prewriter::prewrite
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/txn/commands/prewrite.rs:647:35
 tikv::storage::txn::commands::prewrite::Prewriter::process_write
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/txn/commands/prewrite.rs:535:19
 11: <tikv::storage::txn::commands::prewrite::Prewrite as tikv::storage::txn::commands::WriteCommand<S,L>>::process_write
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/txn/commands/prewrite.rs:250:9
 tikv::storage::txn::commands::Command::process_write
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/txn/commands/mod.rs:691:37
 12: tikv::storage::txn::scheduler::TxnScheduler<E,L>::process_write::{{closure}}::{{closure}}
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1287:21
 tikv::storage::metrics::with_perf_context::{{closure}}
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/metrics.rs:360:19
 std:🧵:local::LocalKey::try_with
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:446:16
 std:🧵:local::LocalKey::with
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:422:9
 tikv::storage::metrics::with_perf_context
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/metrics.rs:351:5
 tikv::storage::txn::scheduler::TxnScheduler<E,L>::process_write::{{closure}}
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1286:17
 13: <core::future::from_generator::GenFuture as core::future::future::Future>::poll
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:91:19
 tikv::storage::txn::scheduler::TxnScheduler<E,L>::process::{{closure}}::{{closure}}
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1177:71
 <core::future::from_generator::GenFuture as core::future::future::Future>::poll
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:91:19
 <resource_metering::InTags as core::future::future::Future>::poll
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/resource_metering/src/lib.rs:266:9
 tikv::storage::txn::scheduler::TxnScheduler<E,L>::process::{{closure}}
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1191:9
 <core::future::from_generator::GenFuture as core::future::future::Future>::poll
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:91:19
 tikv::storage::txn::scheduler::TxnScheduler<E,L>::execute::{{closure}}
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/txn/scheduler.rs:748:54
 <core::future::from_generator::GenFuture as core::future::future::Future>::poll
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:91:19
 14: <tracker::tls::TrackedFuture as core::future::future::Future>::poll::{{closure}}
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tracker/src/tls.rs:64:23
 std:🧵:local::LocalKey::try_with
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:446:16
 std:🧵:local::LocalKey::with
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:422:9
 <tracker::tls::TrackedFuture as core::future::future::Future>::poll
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tracker/src/tls.rs:62:9
 tikv_util::yatp_pool::future_pool::PoolInner::spawn::{{closure}}
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/yatp_pool/future_pool.rs:178:27
 <core::future::from_generator::GenFuture as core::future::future::Future>::poll
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:91:19
 yatp::task::future::RawTask::poll
 at rust/git/checkouts/yatp-e704b73c3ee279b6/5572a78/src/task/future.rs:59:9
 15: yatp::task::future::TaskCell::poll
 at rust/git/checkouts/yatp-e704b73c3ee279b6/5572a78/src/task/future.rs:103:9
 <yatp::task::future::Runner as yatp::pool:🏃:Runner>::handle
 at rust/git/checkouts/yatp-e704b73c3ee279b6/5572a78/src/task/future.rs:387:20
 16: <tikv_util::yatp_pool::YatpPoolRunner as yatp::pool:🏃:Runner>::handle
 at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/yatp_pool/mod.rs:199:24
 17: yatp::pool::worker::WorkerThread<T,R>::run
 at rust/git/checkouts/yatp-e704b73c3ee279b6/5572a78/src/pool/worker.rs:48:13
 yatp::pool::builder::LazyBuilder::build::{{closure}}
 at rust/git/checkouts/yatp-e704b73c3ee279b6/5572a78/src/pool/builder.rs:114:25
 std::sys_common::backtrace::rust_begin_short_backtrace
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:121:18
 18: std:🧵:Builder::spawn_unchecked::{{closure}}::{{closure}}
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:551:17
 <core::panic::unwind_safe::AssertUnwindSafe as core::ops::function::FnOnce<()>>::call_once
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:271:9
 std::panicking::try::do_call
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:483:40
 std::panicking::try
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:447:19
 std::panic::catch_unwind
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:137:14
 std:🧵:Builder::spawn_unchecked::{{closure}}
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:550:30
 core::ops::function::FnOnce::call_once{{vtable.shim}}
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:513:5
 19: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce>::call_once
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2000:9
 <alloc::boxed::Box<F,A> as core::ops::function::FnOnce>::call_once
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2000:9
 std::sys::unix:🧵:Thread:🆕:thread_start
 at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/thread.rs:108:17
 20: start_thread
 21: __clone3
"] [location=components/txn_types/src/types.rs:611] [thread_name=sched-worker-pool-1] [thread_id=0x5]"}

```

### 4. What is your TiDB version? (Required)

90628349860

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #48804

Problem Summary:

When `ResetEngine` failed because of timestamp allocation error like `PD:tso:ErrGenerateTimestamp`, the engine manager is not cleaned. In the next retry, engine manager invokes `ImportEngine` again, leaving `backend.TS` as the zero value.

https://github.com/pingcap/tidb/blob/c4677188c43e0a057d4ea28a7615ebbd9f5fe745/br/pkg/lightning/backend/local/local.go#L1761-L1763

```

[2023/11/21 22:19:52.937 +08:00] [ERROR] [backend.go:222] ["ingest data into storage error"] ["index ID"=15] ["usage info"="disk usage: 100037206016/1968786661376, backend usage: 0"]
[2023/11/21 22:19:52.937 +08:00] [ERROR] [pd.go:236] ["updateTS error"] [txnScope=global] [error="rpc error: code = Unknown desc = [PD:tso:ErrGenerateTimestamp]generate timestamp failed, requested pd is not leader of cluster"]
```

### What is changed and how it works?

This PR sets both `engineInfo.openEngine` and `engineInfo.closedEngine` to `nil` if there is an error during `ResetEngine`. In the next retry, engine manager will open a new backend and allocate TS.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/local/local.go [hunk 1]
```diff
@@ -1761,12 +1761,28 @@ func (local *Backend) ResetEngine(ctx context.Context, engineUUID uuid.UUID) err
...
+			localEngine.TS = 0
+			failpoint.Return(errors.Trace(&mockGRPCErr{}))
+		})
 	}
 	localEngine.pendingFileSize.Store(0)
 
 	return err
 }
 
+type mockGRPCErr struct {
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/local/local.go:1775]: `status.FromError` can be used to convert any err that's not an gRPC error to an UNKNOWN  grpc error

File: pkg/ddl/ingest/backend.go [hunk 2]
```diff
@@ -214,18 +215,46 @@ func (bc *litBackendCtx) Flush(indexID int64, mode FlushMode) (flushed, imported
...
 		return true, false, err
 	}
 	return true, true, nil
 }
 
+func (bc *litBackendCtx) unsafeImportAndReset(ei *engineInfo) error {
+	logutil.Logger(bc.ctx).Info(LitInfoUnsafeImport, zap.Int64("index ID", ei.indexID),
+		zap.String("usage info", bc.diskRoot.UsageInfo()))
+	logger := log.FromContext(bc.ctx).With(
+		zap.String("engineTag", "<import-and-reset>"),
```
Comment (D3Hunter) [inline pkg/ddl/ingest/backend.go:229]: value is fixed, do we need this ?

**Review by wjhuang2016 — APPROVED** [review]

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: tests/realtikvtest/addindextest1/disttask_test.go [hunk 3]
```diff
@@ -229,3 +233,54 @@ func TestAddIndexForCurrentTimestampColumn(t *testing.T) {
...
+	dts := []types.Datum{types.NewIntDatum(1)}
+	sctx := tk.Session().GetSessionVars().StmtCtx
+	idxKey, _, err := tablecodec.GenIndexKey(sctx, tblInfo, idxInfo, tblInfo.ID, dts, kv.IntHandle(1), nil)
+	require.NoError(t, err)
+
+	tikvStore := dom.Store().(helper.Storage)
+	newHelper := helper.NewHelper(tikvStore)
+	mvccResp, err := newHelper.GetMvccByEncodedKeyWithTS(idxKey, 0)
+	require.NoError(t, err)
+	if mvccResp == nil || mvccResp.Info == nil || len(mvccResp.Info.Writes) == 0 {
```
Comment (D3Hunter) [inline tests/realtikvtest/addindextest1/disttask_test.go:281]: require.NotNil(t, mvccResp)

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: br/pkg/lightning/backend/local/local.go (hunk 1 above)
Comment (tangenta) [reply br/pkg/lightning/backend/local/local.go:1775]: cannot use mockGRPCErr (variable of type *"google.golang.org/grpc/internal/status".Status) as error value in argument to errors.Trace: *"google.golang.org/grpc/internal/status".Status does not implement error (missing method Error)compiler[InvalidIfaceAssign](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#InvalidIfaceAssign)

**Pull request merged**